numpy/core/src/multiarray/convert.c

```diff
@@ -321,37 +329,54 @@ PyArray_ToString(PyArrayObject *self, NPY_ORDER order)
         ret = PyBytes_FromStringAndSize(PyArray_DATA(self), (Py_ssize_t) numbytes);
     }
     else {
-        PyObject *new;
-        if (order == NPY_FORTRANORDER) {
-            /* iterators are always in C-order */
-            new = PyArray_Transpose(self, NULL);
-            if (new == NULL) {
-                return NULL;
-            }
-        }
-        else {
-            Py_INCREF(self);
-            new = (PyObject *)self;
-        }
-        it = (PyArrayIterObject *)PyArray_IterNew(new);
-        Py_DECREF(new);
-        if (it == NULL) {
+        iter = NpyIter_New(self, NPY_ITER_READONLY |
+                                NPY_ITER_GROWINNER |
+                                NPY_ITER_EXTERNAL_LOOP |
+                                NPY_ITER_DONT_NEGATE_STRIDES,
+                           order, NPY_NO_CASTING, NULL);
+        if (iter == NULL) {
             return NULL;
         }
+        NPY_BEGIN_THREADS_DEF;
+        needs_api = NpyIter_IterationNeedsAPI(iter);
```

Comment:
i was a bit confused about the needs api stuff and the macros for NPY_BEGIN_THREADS_DEF , NPY_BEGIN_THREADS_NDITER etc. My understanding is that we just want to hold the GIL and I  wasn't sure if we need any of this stuff.